<a href="https://colab.research.google.com/github/Neersha/Python/blob/main/basic_python3.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**AUTHOR**: SOHAM NAHA  
**GITHUB**: www.github.com/soham2109

Python's Features
=================
**Features of Python**

-   Designed by Guido van Rossum in the 1980s-90s.

-   Open-source, interpreted, high-level, general-purpose
-   Dynamically-typed.
-   Supports multiple programming language paradigms
-   Often described as a "batteries included" language due to its comprehensive standard library.
-   Great online community
-   Python was named after the BBC show \"Monty Python's Flying Circus\"\
-   Ranks $3^{rd}$ in [TIOBE's index](https://www.tiobe.com/tiobe-index/) and $1^{st}$ in [PYPL ranking](https://pypl.github.io/PYPL.html).

Journey into Python
===================

Comparison of code-lengths **Hello World! program in the top 3
programming languages.**


C++

```cpp
#include<iostream>
using namespace std;

main(){
	cout << "Hello World." << endl;
}
```

Java

```java
class HelloWorld {
    public static void main(String[] args) {
        System.out.println("Hello World.");
    }
}
```

Python

```python
print("Hello World.")
```


### Hello World in Python
First thing that we want to learn is the `print` command and how to comment somethings out.  
```python
# Single Line Comment
# Used mostly to describe what a line of code does
"""
Multi-line comment in three quotes

Used a DOCSTRINGS to describe what a
function or class performs.
"""
print("Hello World")
```

We can say that """ ... """ is a multi-line comment but not truly. It is actually a multi-line string, and as it is not assigned to any variable it is ignored.

In [ ]:
"""
Multi-line comment in three quotes
Usually used to describe what a function or
class performs.
"""

'\nMulti-line comment in three quotes\nUsually used to describe what a function or \nclass performs.\n'

In [ ]:
print("Hello", "IITB") # Prints Hello World

Hello IITB


Python's Philosophy **The Zen of Python**

In [ ]:
import this

The Zen of Python, by Tim Peters

Beautiful is better than ugly.
Explicit is better than implicit.
Simple is better than complex.
Complex is better than complicated.
Flat is better than nested.
Sparse is better than dense.
Readability counts.
Special cases aren't special enough to break the rules.
Although practicality beats purity.
Errors should never pass silently.
Unless explicitly silenced.
In the face of ambiguity, refuse the temptation to guess.
There should be one-- and preferably only one --obvious way to do it.
Although that way may not be obvious at first unless you're Dutch.
Now is better than never.
Although never is often better than *right* now.
If the implementation is hard to explain, it's a bad idea.
If the implementation is easy to explain, it may be a good idea.
Namespaces are one honking great idea -- let's do more of those!


### Help for what a command does?
Python has a built-in tool called `help()` that allows users to infer or query the things the syntax, inputs, outputs, etc. that a function or command might perform.  
To use: e.g. `help(print)`

In [ ]:
help(print)

Help on built-in function print in module builtins:

print(...)
    print(value, ..., sep=' ', end='\n', file=sys.stdout, flush=False)
    
    Prints the values to a stream, or to sys.stdout by default.
    Optional keyword arguments:
    file:  a file-like object (stream); defaults to the current sys.stdout.
    sep:   string inserted between values, default a space.
    end:   string appended after the last value, default a newline.
    flush: whether to forcibly flush the stream.



In [ ]:
class color:
   PURPLE = '\033[95m'
   BLUE = '\033[96m'
   DARKBLUE = '\033[36m'
   BLUE = '\033[94m'
   GREEN = '\033[92m'
   YELLOW = '\033[93m'
   RED = '\033[91m'
   BOLD = '\033[1m'
   UNDERLINE = '\033[4m'
   END = '\033[0m'

### Data-types in Python
Python supports multiple data-types. They are:  

- integers `int` , e.g. 5
- floating point numbers `float` , e.g. 3.0
- strings `str`, e.g. "IITB"
- boolean `bool`, e.g. True
- complex `complex` , e.g. 2+3j

How to check the type of a variable?   
Two possible ways:

- First, let python decide the type for you using the `type()` command in python.
- Else, check ourselves which data-type instance does the value belong to, using the `isinstance()` command

In [ ]:
# USING type() command
print("------------------")
print("Using type() command")
print("------------------")
print(color.BLUE,"type(5): ",color.END, type(5))
print(color.BLUE,"type(3.0): ",color.END, type(3.0))
print(color.BLUE,"type('IITB'): ",color.END, type("IITB"))
print(color.BLUE,"type(True): ",color.END, type(True))
print(color.BLUE,"type(2+3j): ",color.END, type(2+3j))

# USING ISINSTANCE COMMAND
print("\n\n------------------")
print("Using isinstance() command")
print("------------------")
print(color.BLUE,"isinstance(5, int): ",color.END,isinstance(5, int))
print(color.BLUE,"isinstance(5.045, float): ",color.END,isinstance(5.045, float))
print(color.BLUE,"isinstance('5', str): ",color.END,isinstance('5', str))
print(color.BLUE,"isinstance(5+5j, complex): ",color.END,isinstance(5+5j, complex))
print(color.BLUE,"isinstance(False, bool): ",color.END,isinstance(False, bool))
print(color.BLUE,"isinstance(5, bool): ",color.END,isinstance(5, bool))

------------------
Using type() command
------------------
 type(5):   <class 'int'>
 type(3.0):   <class 'float'>
 type('IITB'):   <class 'str'>
 type(True):   <class 'bool'>
 type(2+3j):   <class 'complex'>


------------------
Using isinstance() command
------------------
 isinstance(5, int):   True
 isinstance(5.045, float):   True
 isinstance('5', str):   True
 isinstance(5+5j, complex):   True
 isinstance(False, bool):   True
 isinstance(5, bool):   False


### Python Variables
Variables as the name suggests are those containers or placeholders of values that may vary. Some simple examples are as follows:

In [ ]:
# int x = 5
boolean = True #<bool> data type
x = 5 # <int> data-type
y = 5.5 # <float> data-type
sci_not = 5.5e3 # scientific notation
z = "IIT Bombay" # <str> data-type
a = 'c' # also <str>
b=2+3j # <complex> data-type

print("----------------")
print("Checking the type of variables")
print("----------------")
print(color.BLUE,"Type of boolean: ",color.END,type(boolean))
print(color.BLUE,"Type of x: ",color.END,type(x))
print(color.BLUE,"Type of y: ",color.END,type(y))
print(color.BLUE,"Type of sci_not: ",color.END,type(sci_not))
print(color.BLUE,"Type of z: ",color.END,type(z))
print(color.BLUE,"Type of a: ",color.END,type(a))
print(color.BLUE,"Type of b: ",color.END,type(b))
print(color.BLUE,"Type of b.conjugate(): ",color.END,type(b.conjugate()))
print(color.BLUE,"b: ",color.END,b,color.BLUE,"b.conjugate(): ",color.END,b.conjugate())

----------------
Checking the type of variables
----------------
 Type of boolean:   <class 'bool'>
 Type of x:   <class 'int'>
 Type of y:   <class 'float'>
 Type of sci_not:   <class 'float'>
 Type of z:   <class 'str'>
 Type of a:   <class 'str'>
 Type of b:   <class 'complex'>
 Type of b.conjugate():   <class 'complex'>
 b:   (2+3j)  b.conjugate():   (2-3j)


### Variable Naming Conventions
Source: [PEP 8 Styling](https://www.python.org/dev/peps/pep-0008/)

-   **Readability is important**

-   Descriptive names are very useful.

-   Python keywords cannot be used as variable names.  
    The following are the 35 python keywords:  
    
    `False` `None` `True` `and` `as` `assert` `async` `await` `break` `class` `continue` `def` `del` `elif` `else` `except` `finally` `for` `from` `global` `if` `import` `in` `is` `lambda` `nonlocal` `not` `or` `pass` `raise` `return` `try` `while` `with` `yield `

-   Use a single trailing underscore '\_' with keywords, if needed.

-   Variables names must start with a letter or an underscore.

-   The remainder of your variable name may consist of letters, numbers
    and underscores.

-   Python is case sensitive.

-   Avoid using the lowercase letter 'l', uppercase 'O' and uppercase
    'I'.

-   Can use snake\_case or CamelCase or mixedCase.

In [ ]:
# To find out the python keywords
import keyword
print("Total number of keywords in python: ", len(keyword.kwlist))
print("The keywords are: \n", *keyword.kwlist)

Total number of keywords in python:  35
The keywords are: 
 False None True and as assert async await break class continue def del elif else except finally for from global if import in is lambda nonlocal not or pass raise return try while with yield


In [ ]:
# variable naming Conventions
my_int = 1 #snake_case
MyInt = 2 # CamelCase
myInt = 3 #mixedCase
my_Int_case = 4 #also mix

### Operations in Python
We have different types of operations in python, like in any other programming language that you may come across. They are:  
- Arithmetic Operations (+, -, /, \*, %, //,  $**$)

**Symbol**  |    **Task**
------------| ----------------
    $+$     |     Addition
    $-$     |   Subtraction
    $/$     |     Division
    $*$     |  Multiplication
    %    |      Modulo
    $//$    |  Floor Division
    $**$    |     exponent

In [ ]:
print("---------------------")
print("ARITHMETIC OPERATIONS")
print("--------------------")
# double assignment
x, y = 1+2, 5-1
print("x+y =", x*y, "where x= {} and y= {}".format(x,y)) #prints 12
u, v = "-", "+"
print("u+v = ",u+v, "where u = {} and v = {}".format(u, v))  #prints "-+"
print("(u+v)*3", (u+v)*3, "where (u+v) = '{}'".format(u+v))
#prints "-+-+-+"

#now if we set y = 5.0-1
y = 5.0-1 # typecast to float
print(color.DARKBLUE,"x = ",color.END,x,color.DARKBLUE," || y =",color.END,y,sep="")
print(color.GREEN, "x*y = ", color.END,x*y,sep="") # prints 12.0

#calculate 3^4 = 81
print(color.GREEN, "3**4 = ", color.END,3**4,sep="") # 81
# check modulo 3
print(color.GREEN, "y%3 = ", color.END, y%3,sep="") #prints 1.0
print(color.GREEN, "14//5 = ", color.END, 14//5,sep="") #prints 2
# above 3 operations does
# not work with strings

---------------------
ARITHMETIC OPERATIONS
--------------------
x+y = 12 where x= 3 and y= 4
u+v =  -+ where u = - and v = +
(u+v)*3 -+-+-+ where (u+v) = '-+'
x = 3 || y =4.0
x*y = 12.0
3**4 = 81
y%3 = 1.0
14//5 = 2


In [ ]:
x,y = divmod(14,5)
print(x,y)

2 4


- Logical Operations (`and`, `or` and `not`)

**Symbol**  |   **Task**
------------| ---------------
    `and`     |  AND operation
    `or`      | OR operation
    `not`     |  NOT operation


In [ ]:
print("---------------------")
print("LOGICAL OPERATIONS")
print("--------------------")
print(color.GREEN, "x =", color.END, x)
print(color.GREEN, "x<10 and x>1 =", color.END,x<10 and x>1) #False
print(color.GREEN, "x<10 or x>5 =", color.END,x<10 or x>5) #True
print(color.GREEN, "not(x<10 and x>5) =", color.END, not(x<10 and x>5)) #True

---------------------
LOGICAL OPERATIONS
--------------------
 x =  2
 x<10 and x>1 =  True
 x<10 or x>5 =  True
 not(x<10 and x>5) =  True



- Relational Operations (==, !=, >, <, >=, <=)  

**Symbol** |  **Task**  
------- |---------------------------
$==$     |        Equivalence
$!=$     |      Not Equivalence
$<$      |        Lesser than
$>$       |      Greater than
$<=$      |Lesser than or equals to
$>=$      |Greater than or equals to


In [ ]:
print("---------------------")
print("RELATIONAL OPERATIONS")
print("--------------------")
print(color.GREEN, "x,y = ", color.END,x,y)
print(color.GREEN, "(x+y>0) = ", color.END, x+y>0) # prints True
print(color.GREEN, "(x==3) = ", color.END, x==3) #prints True
print(color.GREEN, "z =", color.END,z)
print(color.GREEN, "isinstance(z, int))= ", color.END,isinstance(z, int)) # False

---------------------
RELATIONAL OPERATIONS
--------------------
 x,y =   2 4
 (x+y>0) =   True
 (x==3) =   False
 z =  IIT Bombay
 isinstance(z, int))=   False


- Bit-wise Operations (&, |, ~, ^)

**Symbol**  |  **Task**
------------| -------------
    &    |  Logical AND
    $\mid$  |   Logical OR
    $\sim$  |   Logical NOT
$\wedge$    |Logical XOR
    $<<$    |  Left Shift
    $>>$    |  Right Shift


In [ ]:
print("---------------------")
print("BITWISE OPERATIONS")
print("--------------------")
a = 2 #10
b = 3 #11
print(color.RED,"a = ",a," and b = ", b,color.END, sep="")
print(color.RED,"(a & b): ",color.END,a & b,sep="") #prints 2
# bin() to convert to binary
print(color.RED,"bin(a & b): ",color.END, bin(a&b), sep="") #prints 0b10

# Right shift operator
print(color.RED,"5>>1: ",color.END, 5>>1,sep="")
# 5 -> 0000 0101 (assuming 8-bits for an integer)
# Right Shifting binary 5 by
# 1 place produce 0000 0010
# which is 2

print(color.RED,"17>>1: ",color.END,17>>1,sep="")

print(color.RED,"5<<1: ",color.END,5<<1,sep="")
# 5 -> 0000 0101
# Left Shifting binary 5 by
# 1 place produce 0000 1010
# which is 10
print(color.RED,"17<<1: ",color.END,17<<1,sep="")

---------------------
BITWISE OPERATIONS
--------------------
a = 2 and b = 3
(a & b): 2
bin(a & b): 0b10
5>>1: 2
17>>1: 8
5<<1: 10
17<<1: 34


# Python Data-Structures
---
Other than the conventional data-types of python, it also provides many nifty data-structures that comes in very handy.  
The data-structures that we are talking about are as follows:  
- **List**
- **Tuples**
- **Sets**
- **Dictionaries**

## Python Lists
---
- Python lists can be thought of as simple linked-lists that one may have covered in basic data-structures courses, but is a bit more optimized for performance and made easy for us.  
- Python lists can contain any type of data that we want, and can contain mixed types as well
- Python Lists are basically used for storing sequences
- **Lists are mutable**
- They are 0-indexed.
- It offers several options or features to  manipulate and access these sequences.  
    - `copy()` : to create a shallow copy of the list
    - `append()` : to add an element to the end of a list
    - `extend()` : to extend the contents of a list
    - `index()` : to find the index of an item in a list
    - `pop()` : to pop the item at a given index of a list
    - `insert()` : to insert an element at a given index in a list
    - `count()` : to get the number of times an element occurs in a list
    - `remove()` : the remove an element from the list
    - `sort()` : to sort the list elements
    - `reverse()` : to reverse the list elements
    - `clear()` : to delete the list contents

Let's look at these options one by one.

In [ ]:
print(*range(1,10, 3))

1 4 7


In [ ]:
# Create a dummy list of numbers from 0 to 20
dummy_list = list(range(0,21))
print(color.BLUE,"dummy_list:\n", color.END, dummy_list,sep="")
print("-"*20)

# suppose we create another list of number from 50 to 70 and we want to append
# these numbers to the end of the dummy_list
# Create the new list.
new_dummy_list = list(range(50,71,4))
print(color.BLUE,"new_dummy_list:\n", color.END, new_dummy_list,sep="")

# Add the dummy_list.
# There are three or more possible ways we can do this.
# I'll show two. But first let's look at this

dummy_list:
[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20]
--------------------
new_dummy_list:
[50, 54, 58, 62, 66, 70]


### A Quick Look into list copy
---
Lists copies are very essential to understand, as it can prove to be fatal if done wrong.

In [ ]:
dummy_list_copy = dummy_list.copy() # This makes a copy of dummy_list
print("Check if both the lists have the same content.")
print(color.BLUE,"dummy_list_copy == dummy_list:", color.END, dummy_list_copy == dummy_list)
print("Check if both are the same instances.")
print(color.BLUE,"dummy_list_copy is dummy_list:", color.END, dummy_list_copy is dummy_list)


dummy_list_direct = dummy_list
print(color.BLUE,"dummy_list_direct is dummy_list:", color.END,dummy_list_direct is dummy_list)

## So any changes done to dummy_list_direct, gets reflected in dummy_list
## but any changes to dummy_list_copy should not get refleced in dummy_list
dummy_list_direct[3] = 3000
print(color.BLUE,"dummy_list_direct:", color.END, dummy_list_direct)
print(color.BLUE,"dummy_list:", color.END, dummy_list)

dummy_list_copy[3]= 3
print(color.BLUE,"dummy_list_copy:",color.END,dummy_list_copy)
print(color.BLUE,"dummy_list:",color.END,dummy_list)

Check if both the lists have the same content.
 dummy_list_copy == dummy_list:  True
Check if both are the same instances.
 dummy_list_copy is dummy_list:  False
 dummy_list_direct is dummy_list:  True
 dummy_list_direct:  [0, 1, 2, 3000, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20]
 dummy_list:  [0, 1, 2, 3000, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20]
 dummy_list_copy:  [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20]
 dummy_list:  [0, 1, 2, 3000, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20]


## Adding new values a list
---
### Three ways
---
- Step 1: Using `append()`, useful when appending elements
- Step 2: Using `extend()`, useful when appending other lists
- Step 3: Using list addition, useful when appending other lists


- Here we informally look in to `for in` loop of python. More like a customized way of using for loops.  
- Mostly, in other languages like `C` or `Java`, while implementing `for` loops, we have to do somthing like :  
<div class="alert alert-block alert-info">
    <pre>for(int i=0; i<10; i++)</pre>
</div>
for creating a for loop.  
- But in python the `for` (`for ... in`) loops work in a simpler manner.
- It can access the elements of an iterator directly and assign them to a variable, that can be used inside the loop.

In [ ]:
# Now lets go back to our case of adding the new_dummy_list values to dummy_list
# Redefining dummy_list here as it got altered
dummy_list = list(range(0,21))
print(color.BLUE,"dummy_list:\n",color.END,dummy_list)

# Step1:
# This takes each element of new_dummy_list and appends to the end of dummy_list
for dummy in new_dummy_list:
    dummy_list.append(dummy)
print(color.BLUE,"Using Step1, dummy_list:\n",color.END, dummy_list)

# Step2: Using the `extend` option of lists
# Redefining dummy_list here as it got altered
dummy_list = list(range(0,21))
dummy_list.extend(new_dummy_list)
print(color.BLUE,"Using Step2, dummy_list:\n",color.END,dummy_list)

# Step3: Using just addition of lists
# Redefining dummy_list here as it got altered
dummy_list = list(range(0,21))
dummy_list += new_dummy_list
print(color.BLUE,"Using Step3, dummy_list:\n",color.END,dummy_list)

 dummy_list:
  [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20]
 Using Step1, dummy_list:
  [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 50, 54, 58, 62, 66, 70]
 Using Step2, dummy_list:
  [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 50, 54, 58, 62, 66, 70]
 Using Step3, dummy_list:
  [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 50, 54, 58, 62, 66, 70]


### Inserting elements at a given position in a list using `insert()`
----
`list.insert(index, element)` :   
- Insert an element at a given index.  
- The first argument is `index`, the position at which to insert.
- The second argument is `element`, the element to insert at position `index`.

In [ ]:
# Suppose we want to insert an element at position lets say, at index 9 of dummy_list
# It does not replace the element at the posn but inserts another one and shifts the rest
# So length of the list increases.
dummy_list = list(range(0,21))
# len(list) can be used to get the length of a list
print(color.BLUE,"Before insert, dummy_list:\n",color.END,dummy_list,"\nLength:", len(dummy_list),sep="")
dummy_list.insert(9, 25)
print(color.BLUE,"After insert, dummy_list:\n",color.END,dummy_list,"\nLength:", len(dummy_list),sep="")


Before insert, dummy_list:
[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20]
Length:21
After insert, dummy_list:
[0, 1, 2, 3, 4, 5, 6, 7, 8, 25, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20]
Length:22


### `list.count(element)` to count number of occurences of the element in the list

In [ ]:
# Suppose we want to count the number of occurences of an element in the list
# Before the count lets actually create a new dummy_list that contains overlapping occurences
dummy_list += list(range(10, 31))
print(color.RED,"Number of occurences of 25 in dummy_list:", color.END, dummy_list.count(25))
print(color.RED,"Number of occurences of 5 in dummy_list:", color.END, dummy_list.count(5))

 Number of occurences of 25 in dummy_list:  2
 Number of occurences of 5 in dummy_list:  1


### Finding the indices of double occurence of the value 25
----


In [ ]:
dummy_list.index(25, 10)

37

In [ ]:
list_ = [1, 3.0, "IITB", 2+3j]
print(list_)

[1, 3.0, 'IITB', (2+3j)]


In [ ]:
dummy_list[-1]

30

In [ ]:
var=[1,2,4]
var.append(6)
new_var = [5,6,7]
var.extend(new_var)
var = var + new_var
print(var)

[1, 2, 4, 6, 5, 6, 7, 5, 6, 7]


In [ ]:
mid = len(var)//2
var.pop(mid)
print(var)

[1, 2, 4, 6, 5, 7, 5, 6, 7]


In [ ]:
index = list()
for i in range(dummy_list.count(25)):
    if len(index)==0:
        index.append(dummy_list.index(25, 0))
    else:
        index.append(dummy_list.index(25, index[-1]+1))
print(index)

[9, 37]


### Using Lists as Stacks
---
- Stacks is a data-structures that follows the LIFO (last-in-first-out) strategy, e.g. stack of copies in a table.
- Python `list` can be used as a stack

In [ ]:
stack = [1,2,3]
print("Initial Stack:",stack)
stack.extend([10,20,130])
print("Stack after adding elements:", stack)
ele = stack.pop()
print("Popped element:", ele)
print("Stack after popping out the last element:", stack)

Initial Stack: [1, 2, 3]
Stack after adding elements: [1, 2, 3, 10, 20, 130]
Popped element: 130
Stack after popping out the last element: [1, 2, 3, 10, 20]


### Using Lists as Queues
---
- Queues is a data-structures that follows the FIFO (first-in-first-out) strategy, e.g. a queue to buy a ticket.
- Python `list` can be used as a queue

In [ ]:
queue = [1,2,3]
print("Initial Queue:",queue)
new_ele = [10,20,130]
for i in reversed(new_ele):
    queue.insert(0,i)
print("Queue after adding elements:", queue)
ele = queue.pop(0)
print("Popped element:", ele)
print("Queue after popping out the last element:", queue)

Initial Queue: [1, 2, 3]
Queue after adding elements: [10, 20, 130, 1, 2, 3]
Popped element: 10
Queue after popping out the last element: [20, 130, 1, 2, 3]


In [ ]:
var = [1,2,34, 50]
print(*reversed(var))

50 34 2 1


### List Comprehensions
---
List comprehensions provide a concise way to create lists. Common applications are to make new lists where each element is the result of some operations applied to each member of another sequence or iterable, or to create a subsequence of those elements that satisfy a certain condition. [Source](https://docs.python.org/3/tutorial/datastructures.html#list-comprehensions)

- Suppose we have a list (a sequence of numbers)
- We want to calculate the square of each of these numbers to form another list.
- There are two ways to do that:  
    - First, the naive way of looping through the elements, and squaring each element in the list, and then appending the squared value to the other list
    - Second, the one-liner precise way of using list comprehensions.

In [ ]:
numbers = list(range(1,11)) # store numbers from 1 to 10
# First Method
squares = []
for i in numbers:
    squares.append(i**2)
print(color.BLUE,"Using Method 1: squares=",color.END, squares)

# Second Method
squares = [i**2 for i in numbers]
print(color.BLUE,"Using Method 2: squares=",color.END, squares)

 Using Method 1: squares=  [1, 4, 9, 16, 25, 36, 49, 64, 81, 100]
 Using Method 2: squares=  [1, 4, 9, 16, 25, 36, 49, 64, 81, 100]


In [ ]:
print(*squares, sep=" || ")

1 || 4 || 9 || 16 || 25 || 36 || 49 || 64 || 81 || 100


### Nested Lists
----
We can also create nested lists, i.e. lists inside a list.

In [ ]:
nested_example = [[i for i in range(10)] for j in range(5)]
print(color.BOLD,"Nested List:",color.END,*nested_example,sep="\n")

# * (splat) operator is used to open the iterable(in this case a list) up,
# i.e. unpack the values in a list
# Let's see an example:
print(color.BLUE,"Range, without unpacking:",color.END, range(4),sep="")
print(color.BLUE,"Range, with splat unpacking:",color.END,*range(4),sep="")

*x, = range(5)
print(color.BLUE,"Splat operator example:\n",color.END,
      color.BLUE,"*x, = range(5) gives x =",color.END, x, sep="")


Nested List:

[0, 1, 2, 3, 4, 5, 6, 7, 8, 9]
[0, 1, 2, 3, 4, 5, 6, 7, 8, 9]
[0, 1, 2, 3, 4, 5, 6, 7, 8, 9]
[0, 1, 2, 3, 4, 5, 6, 7, 8, 9]
[0, 1, 2, 3, 4, 5, 6, 7, 8, 9]
Range, without unpacking:range(0, 4)
Range, with splat unpacking:0123
Splat operator example:
*x, = range(5) gives x =[0, 1, 2, 3, 4]


### List Slicing
----
So far we have only seen numeric lists. Now, we'll see a list with mixed data-types.

In [ ]:
print(1e10)
print(type(1e10))

10000000000.0
<class 'float'>


In [ ]:
print([1,2,3])
print(repr("IITB"))

[1, 2, 3]
'IITB'


In [ ]:
mixed_list = ["IITB", 8.0, 2021, 1e5, "A"]
print("Complete list: ", mixed_list)
print(color.RED,"Let's look at the individual types, looping thorugh the complete list.",color.END, sep="")
for i,j in enumerate(mixed_list):
    print("Index: {}, element: {}, type: {}".format(i,repr(j),type(j)))
print("-"*20)
print("Sliced list: ",mixed_list[:2])
print(color.RED,"Let's look at the individual types, looping through the first two elements only.",color.END, sep="")
for i,j in enumerate(mixed_list[0:2]):
    print("Index: {}, element: {}, type: {}".format(i,repr(j),type(j)))
print("-"*20)

print("Sliced list (evens only): ",mixed_list[ 0: 6: 3])
print("Sliced list (odds only): ",mixed_list[ 1: : 2])

Complete list:  ['IITB', 8.0, 2021, 100000.0, 'A']
Let's look at the individual types, looping thorugh the complete list.
Index: 0, element: 'IITB', type: <class 'str'>
Index: 1, element: 8.0, type: <class 'float'>
Index: 2, element: 2021, type: <class 'int'>
Index: 3, element: 100000.0, type: <class 'float'>
Index: 4, element: 'A', type: <class 'str'>
--------------------
Sliced list:  ['IITB', 8.0]
Let's look at the individual types, looping through the first two elements only.
Index: 0, element: 'IITB', type: <class 'str'>
Index: 1, element: 8.0, type: <class 'float'>
--------------------
Sliced list (evens only):  ['IITB', 100000.0]
Sliced list (odds only):  [8.0, 100000.0]


## Fibonacci Sequences
---
- Lets now have a small code to demonstrate Fibonacci Squences.  
- In mathematics, the Fibonacci numbers, commonly denoted Fn, form a sequence, called the Fibonacci sequence, such that each number is the sum of the two preceding ones, starting from 0 and 1. [Source](https://en.wikipedia.org/wiki/Fibonacci_number)

> so, $F_0$ = 0, $F_1$ = 1  
and $F_{n}$ = $F_{n-1}$ + $F_{n-2}$.  

- The beginning of the sequence is thus:  
0,1,1,2,3,5,8,13,21,34,55,89,144, $\ldots$  
- Lets look at the code for this and reveal some interesting facts.
- **Trivia**  
In mathematics, two quantities are in the golden ratio ($\phi$) (something around ~1.61803398874) if their ratio is the same as the ratio of their sum to the larger of the two quantities. [Source](https://en.wikipedia.org/wiki/Golden_ratio)

In [ ]:
# let's define upto how many numbers do we want the sequence to be
n=20
# Initialize the list with values 0,1
fib_seq = [0,1]
# now we iterate and loop upto n and append values to the sequence
for i in range(len(fib_seq),n):
    fib_seq.append(fib_seq[-1]+fib_seq[-2])

print(color.BLUE,"Fibonacci Sequence upto {} numbers is:\n".format(n),color.END, fib_seq)

# Now, let's look into an interesting fact
# The ratio of consecutive numbers, especially starting from
# index 2, actually converges to a number of mathematical
# significance, the Golden Ratio
fib_ratio = [fib_seq[i]/fib_seq[i-1] for i in range(len(fib_seq)) if i>2]
print(color.BLUE,"Ratios are:\n".format(n),color.END, fib_ratio)

 Fibonacci Sequence upto 20 numbers is:
  [0, 1, 1, 2, 3, 5, 8, 13, 21, 34, 55, 89, 144, 233, 377, 610, 987, 1597, 2584, 4181]
 Ratios are:
  [2.0, 1.5, 1.6666666666666667, 1.6, 1.625, 1.6153846153846154, 1.619047619047619, 1.6176470588235294, 1.6181818181818182, 1.6179775280898876, 1.6180555555555556, 1.6180257510729614, 1.6180371352785146, 1.618032786885246, 1.618034447821682, 1.6180338134001253, 1.618034055727554]


In [ ]:
print("Min:",min(fib_ratio))
print("Max:",max(fib_ratio))
print("Mean of the fib_ratios= {:.3f}".format(sum(fib_ratio)/len(fib_ratio)))

Min: 1.5
Max: 2.0
Mean of the fib_ratios= 1.636


## Tuples
----
- Tuples is another Python data-structure for holding sequential heterogenous data.
- It's difference with `list` type is that **`tuples` are immutable**.
- We saw that `lists` are always defined using `[]` brackets. But `tuples` are defined using `()`.
- A special problem is the construction of tuples containing 0 or 1 items: the syntax has some extra quirks to accommodate these.
- Empty tuples are constructed by an empty pair of parentheses
- A tuple with one item is constructed by following a value with a comma (it is not sufficient to enclose a single value in parentheses).

In [ ]:
# way to initialize empty tuple
empty_tuple = ()
print(color.GREEN,"type(empty_tuple): ",color.END,type(empty_tuple),sep="")
# way to initialize single element tuple
# if comma is not provided turns into a generator object
single_element_tuple = (1,)
print(color.GREEN,"type(single_element_tuple): ",color.END,type(single_element_tuple),
      " where single_element_tuple=(1,)", sep="")
single_element_tuple = (1)
print(color.GREEN,"type(single_element_tuple): ",color.END,type(single_element_tuple),
      " where single_element_tuple=(1)",sep="")

# this is a tuple
example_tuple = (1,2,3,4,10,100,2,33,"IITB","A",55.023)
# this is also a tuple
example_tuple2 = 1,2,3,4,10,100,2,33,"IITB","A",55.023
print("example_tuple: ",example_tuple)
print("type of example_tuple2: ",type(example_tuple2))

print(color.RED,"index of 'IITB' in example_tuple",color.END, example_tuple.index("IITB"))
print(color.RED,"Number of occurences of 2 in example_tuple",color.END, example_tuple.count(2))
example_tuple += (35,20,"Add")
print(color.RED,"example_tuple after adding elements",color.END,example_tuple)
print(color.RED,"example_tuple[10]:",color.END, example_tuple[10])
print(color.RED,"#elements example_tuple",color.END, len(example_tuple))
print(color.RED,"Tuple Slicing, example_tuple[2:6]",color.END, example_tuple[2:6])

type(empty_tuple): <class 'tuple'>
type(single_element_tuple): <class 'tuple'> where single_element_tuple=(1,)
type(single_element_tuple): <class 'int'> where single_element_tuple=(1)
example_tuple:  (1, 2, 3, 4, 10, 100, 2, 33, 'IITB', 'A', 55.023)
type of example_tuple2:  <class 'tuple'>
 index of 'IITB' in example_tuple  8
 Number of occurences of 2 in example_tuple  2
 example_tuple after adding elements  (1, 2, 3, 4, 10, 100, 2, 33, 'IITB', 'A', 55.023, 35, 20, 'Add')
 example_tuple[10]:  55.023
 #elements example_tuple  14
 Tuple Slicing, example_tuple[2:6]  (3, 4, 10, 100)


TypeError: ignored

## Python Sets
---
- This data-structure is unordered in contrast to `lists` and `tuples`.
- As with sets in mathematics that only contain unique elements, python `sets` also contain no duplicate element.
- Set objects support mathematical operations like `union`, `intersection`, `difference`, and `symmetric difference`.
- Curly braces or the set() function can be used to create sets.
- But, **empty set are initialized using `set()`**, because `{}` creates an empty `dictionary` another one of python's built-in data-structures.
- Similar to list comprehensions that we saw a while ago, set also supports similar comprehensions
- `set.add()` is used to add elements to the set as `append()` won't work for sets.

> Set Operations:
>>`key in s`         # containment check  
`key not in s`   # non-containment check  
`s1 == s2`       # s1 is equivalent to s2  
`s1 != s2`       # s1 is not equivalent to s2  
`s1 <= s2`    # s1is subset of s2   
`s1 < s2` or `s1.issubset(s2)`     # s1 is proper subset of s2   
`s1 >= s2`             # s1 is superset of s2  
`s1 > s2` or `s1.issuperset(s2)`    # s1 is proper superset of s2  
`s1 | s2` or `s1.union(s2)`       # the union of s1 and s2  
`s1 & s2` or `s1.intersection(s2)`        # the intersection of s1 and s2   
`s1 – s2` or `s1.difference(s2)`        # the set of elements in s1 but not s2   
`s1 ˆ s2` or `s1.symmetric_difference(s2)`         # the set of elements in precisely none of s1 or s2  

In [ ]:
empty_set = set()
empty_set2 = {}
print(color.BLUE,"Type of `set()` is: ",color.END,type(empty_set),sep="")
print(color.BLUE,"While type of `{}` is: ",color.END,type(empty_set2),sep="")

# Lets create a set of fruit in a basket
basket = {'apple', 'orange', 'apple', 'pear', 'orange', 'banana'}
print(color.BLUE,"Basket contains: ",color.END, basket, sep="")
print(color.BLUE,"Number of unique fruits in the basket: ",color.END, len(basket), sep="")
basket.add("watermelon")
print(color.BLUE,"Now, Basket contains: ",color.END, basket, sep="")

Type of `set()` is: <class 'set'>
While type of `{}` is: <class 'dict'>
Basket contains: {'orange', 'pear', 'apple', 'banana'}
Number of unique fruits in the basket: 4
Now, Basket contains: {'orange', 'pear', 'apple', 'banana', 'watermelon'}


In [ ]:
set1 = set(range(6))
print(color.GREEN,"set1: ",color.END,set1)
set2 = set(range(3,10))
print(color.GREEN,"set2: ",color.END,set2)

print(color.GREEN,"set1.union(set2): ",color.END, set1.union(set2))
print(color.GREEN,"set1.intersection(set2): ",color.END, set1.intersection(set2))
print(color.GREEN,"set1.difference(set2): ", color.END, set1.difference(set2))
print(color.GREEN,"set1.symmetric_difference(set2): ",color.END,set1.symmetric_difference(set2))

print(color.GREEN,"set1.isdisjoint(set2): ",color.END, set1.isdisjoint(set2))
print(color.GREEN,"set1.isdisjoint({10,12,13}): ",color.END, set1.isdisjoint({10,12,13}))

print(color.GREEN,"Is set1 a superset of set2 (set1 > set2): ",color.END, set1>set2)

 set1:   {0, 1, 2, 3, 4, 5}
 set2:   {3, 4, 5, 6, 7, 8, 9}
 set1.union(set2):   {0, 1, 2, 3, 4, 5, 6, 7, 8, 9}
 set1.intersection(set2):   {3, 4, 5}
 set1.difference(set2):   {0, 1, 2}
 set1.symmetric_difference(set2):   {0, 1, 2, 6, 7, 8, 9}
 set1.isdisjoint(set2):   False
 set1.isdisjoint({10,12,13}):   True
 Is set1 a superset of set2 (set1 > set2):   False


In [ ]:
{1,2,3,4,5}>{1,2,3}

True

## Python Dictionaries
---
- Dictionaries in python are another type of data-structures that work on the concept of **key-value pairs**, i.e. more like associative arrays.
- Each key-value pair maps the key to its associated value.
- Dictionaries are **mutable**
- Can be nested.
- Dictionaries are defined using `{}`, like `sets`. But unlike `sets`, dictionaries have to be defined in the form of `{"key": <value>}`.
- Dictionaries in Python are maintained in an ordered fashion.
> Dictionary Operations:  
>> `dict.keys()` to obtain the keys of the dictionary.   
`dict.values()` to obtain the values of the dictionary.  
The `keys` cannot be something that is mutable, like lists, etc.  
`dict.clear()` clears the dictionary  
`dict.get(key)`  gets the values corresponding to key `key`  
`dict.items()` returns a list of key-value pairs, with the first value of the pair being the key and the second being the value correponding to the key  
`dict.pop(key)` removes the values corresponding to the key `key`  
`dict.popitem()` pops the last key-value pair added to the dictionary and returns the popped item.
`dict.update(<obj>)` merges the two dictionary objects  


In [ ]:
empty_dict = {}
print(color.BLUE,"type({}): ",color.END, type(empty_dict), sep="")

# Here we create a dummy IPL team details dictionary
# Data borrowed from iplt20.com
ipl_dict = {"IPL Teams": ["MI","CSK","KKR","RCB", "RR", "KXIP", "SRH", "DD"],
                "Captains": ["Rohit Sharma", "Mahendra Singh Dhoni", "Eoin Morgan",
                             "Virat Kohli", "Sanju Samson", "KL Rahul", "David Warner",
                             "Shreyas Iyer"],
                "Coach": ["Mahela Jayawardene", "Stephen Fleming", "Brendon McCullum",
                          "Simon Katich", "Andrew McDonald", "Anil Kumble", "Trevor Bayliss",
                          "Ricky Ponting"]}

print(color.BOLD,"The dictionary keys: ",color.END, ipl_dict.keys(), sep="")
print(color.BOLD,"The dictionary values:\n",color.END, ipl_dict.values(), sep="")

# another way to obtain the key-value pairs, is by using dict.items()
# dict.items() returns tuples, so we can do something like the following
for key,value in ipl_dict.items():
    print(color.BOLD,"key: ",color.END, color.BLUE,key,color.END, color.BOLD, " value: ",
          color.END, color.BLUE,value,color.END, sep="")

type({}): <class 'dict'>
The dictionary keys: dict_keys(['IPL Teams', 'Captains', 'Coach'])
The dictionary values:
dict_values([['MI', 'CSK', 'KKR', 'RCB', 'RR', 'KXIP', 'SRH', 'DD'], ['Rohit Sharma', 'Mahendra Singh Dhoni', 'Eoin Morgan', 'Virat Kohli', 'Sanju Samson', 'KL Rahul', 'David Warner', 'Shreyas Iyer'], ['Mahela Jayawardene', 'Stephen Fleming', 'Brendon McCullum', 'Simon Katich', 'Andrew McDonald', 'Anil Kumble', 'Trevor Bayliss', 'Ricky Ponting']])
key: IPL Teams value: ['MI', 'CSK', 'KKR', 'RCB', 'RR', 'KXIP', 'SRH', 'DD']
key: Captains value: ['Rohit Sharma', 'Mahendra Singh Dhoni', 'Eoin Morgan', 'Virat Kohli', 'Sanju Samson', 'KL Rahul', 'David Warner', 'Shreyas Iyer']
key: Coach value: ['Mahela Jayawardene', 'Stephen Fleming', 'Brendon McCullum', 'Simon Katich', 'Andrew McDonald', 'Anil Kumble', 'Trevor Bayliss', 'Ricky Ponting']


In [ ]:
ipl_stadium_dict={"stadiums": ["Wankhede Stadium", "M. A. Chidambaram Stadium", "Eden Gardens",
                               "M. Chinnaswamy Stadium", "Sawai Mansingh Stadium",
                               "IS Bindra Stadium", "Rajiv Gandhi Intl. Cricket Stadium",
                               "Arun Jaitley Stadium"]}

ipl_dict.update(ipl_stadium_dict)
for key,value in ipl_dict.items():
    print(color.BOLD,"key: ",color.END, color.BLUE,key,color.END, color.BOLD, " value: ",
          color.END, color.BLUE,value,color.END, sep="")

key: IPL Teams value: ['MI', 'CSK', 'KKR', 'RCB', 'RR', 'KXIP', 'SRH', 'DD']
key: Captains value: ['Rohit Sharma', 'Mahendra Singh Dhoni', 'Eoin Morgan', 'Virat Kohli', 'Sanju Samson', 'KL Rahul', 'David Warner', 'Shreyas Iyer']
key: Coach value: ['Mahela Jayawardene', 'Stephen Fleming', 'Brendon McCullum', 'Simon Katich', 'Andrew McDonald', 'Anil Kumble', 'Trevor Bayliss', 'Ricky Ponting']
key: stadiums value: ['Wankhede Stadium', 'M. A. Chidambaram Stadium', 'Eden Gardens', 'M. Chinnaswamy Stadium', 'Sawai Mansingh Stadium', 'IS Bindra Stadium', 'Rajiv Gandhi Intl. Cricket Stadium', 'Arun Jaitley Stadium']


In [ ]:
dict_ = {**ipl_dict}
print(dict_)

{'IPL Teams': ['MI', 'CSK', 'KKR', 'RCB', 'RR', 'KXIP', 'SRH', 'DD'], 'Captains': ['Rohit Sharma', 'Mahendra Singh Dhoni', 'Eoin Morgan', 'Virat Kohli', 'Sanju Samson', 'KL Rahul', 'David Warner', 'Shreyas Iyer'], 'Coach': ['Mahela Jayawardene', 'Stephen Fleming', 'Brendon McCullum', 'Simon Katich', 'Andrew McDonald', 'Anil Kumble', 'Trevor Bayliss', 'Ricky Ponting'], 'stadiums': ['Wankhede Stadium', 'M. A. Chidambaram Stadium', 'Eden Gardens', 'M. Chinnaswamy Stadium', 'Sawai Mansingh Stadium', 'IS Bindra Stadium', 'Rajiv Gandhi Intl. Cricket Stadium', 'Arun Jaitley Stadium']}


In [ ]:
# Another one-liner way to merge-dictionaries is as follows
dummy_dict={**ipl_dict, **ipl_stadium_dict}
print(color.BOLD,"The dictionary keys of dummy_dict are: ",color.END,dummy_dict.keys(), sep="")
k,v = dummy_dict.popitem()
print("Popped item:\nKey: ",k,"\nValue: ",v,sep="")
print("-"*20)
print()

dummy_dict.pop("Coach")
print(color.BOLD,"The dictionary keys after popping `Coach` are: ",color.END, dummy_dict.keys(), sep="")
print("-"*20)
print()
print(color.BOLD,"dummy_dict.get(`Captains`): ",color.END,dummy_dict.get("Captains"), sep="")

The dictionary keys of dummy_dict are: dict_keys(['IPL Teams', 'Captains', 'Coach', 'stadiums'])
Popped item:
Key: stadiums
Value: ['Wankhede Stadium', 'M. A. Chidambaram Stadium', 'Eden Gardens', 'M. Chinnaswamy Stadium', 'Sawai Mansingh Stadium', 'IS Bindra Stadium', 'Rajiv Gandhi Intl. Cricket Stadium', 'Arun Jaitley Stadium']
--------------------

The dictionary keys after popping `Coach` are: dict_keys(['IPL Teams', 'Captains'])
--------------------

dummy_dict.get(`Captains`): ['Rohit Sharma', 'Mahendra Singh Dhoni', 'Eoin Morgan', 'Virat Kohli', 'Sanju Samson', 'KL Rahul', 'David Warner', 'Shreyas Iyer']


### Interesting thing to note about dictionaries
---

In [ ]:
dict_ = {(1,1): "2", (2,2): "4"}
print(dict_)

# dict_ = {[1,1]: "2", [2,2]: "4"}
# print(dict_)

try :
    dict_ = {[1,1]: "2", [2,2]: "4"}
    print(dict_)
except Exception as e:
    print("Error: {}".format(e))

{(1, 1): '2', (2, 2): '4'}


TypeError: ignored

- Hashing is a way to convert one value into another of a fixed size, that can be used as a unique number.
- In python, dictionary keys are stored as hash values internally.
- So, the keys that are provided to a dictionary has to be hashable.
- In python, there is a command called `hash()` that can be used to hash a value. Let's see some examples.

In [ ]:
print(color.RED,"hash('Hello'): ",color.END,hash("Hello"),sep="")
print(color.RED,"hash(25): ",color.END,hash(25),sep="")
print(color.RED,"hash((1,2,3,4)): ",color.END,hash((1,2,3,4)),sep="")
try:
    print(color.RED,"hash([1,2,3,4]): ",color.END,hash([1,2,3,4]),sep="")
except Exception as e:
    print(color.RED,"Error: ",color.END,e, sep="")

hash('Hello'): -2462362361584402639
hash(25): 25
hash((1,2,3,4)): 485696759010151909
Error: unhashable type: 'list'


## Python Strings

- Strings in python can be thought of as lists of characters.
- We can access each element of the string separately, although we **cannot change the value at the index**.
- Python provides a lot of functionalities for string processing, like:   
    - `lower()`
    - `upper()`
    - `title()`
    - `swapcase()`
    - `capitalize()`
    - `capwords()`
    - `startswith()`
    - `endswith()`
    - `format()`
    - `join()`
    - `strip()`
    - `split()`
    - `index()`

- Python also provides a built-in `string` module, that contains additional functionalitites for string processing.

#### Let's look at some of these features

In [ ]:
# Example dummy string, the famous Lorem Ipsum
import string


# gives all the punctuation characters
punctuations = string.punctuation
print(color.BOLD + "Punctuation: " + color.END + punctuations)
print("-"*20)

# single character, also a string in python
char_string = 'a'
print("Type of char_string: {}".format(type(char_string)))
print("-"*20)

# word-string
word_string="Hello"
print("Character at position {} of word_string is : '{}'".format(3, word_string[3]))
print("-"*20)

# single-line string
single_line = "Hello World."
print("Single Line string: ",single_line)
print("-"*20)

# Below is a multi-line string. each line ends with a "\n" character
lorem_ipsum_string="""
Lorem ipsum dolor sit amet, consectetur adipiscing elit.
Duis laoreet tempor tempus.
Nulla vel mi vel nibh congue interdum.
Pellentesque elit nisl, vulputate quis erat eget, eleifend maximus lorem.
Praesent placerat sem in ipsum ultrices, sed tincidunt quam consectetur.
Donec vel nisi et lorem porta posuere vitae vitae justo.
Pellentesque rhoncus scelerisque ex, facilisis consequat leo aliquam vel.
Vestibulum facilisis, nibh sit amet luctus dapibus, tellus magna euismod nisi, in condimentum mi odio sed ligula.
Aliquam purus est, imperdiet ac porttitor ac, eleifend in odio."""

# to get sentences first strip off extra spaces that might be present
# at the beginning or end of the dummy text and then split with respect to
# the character "\n" which is the new-line character.
sentences = lorem_ipsum_string.strip().split("\n")

# and empty placeholder where the words are to be stored
words = list()

# Access each elements in the sentences list
# and then split them according to spaces and other punctuations
for sentence in sentences:
    print("-"*20)
    print(color.RED,"Sentence: ",color.END,sentence)
    # see if the characters are present in the punctuations list
    # if they are not present add them to the dummy list
    # after passing throught the whole sentence, and removing all the
    # punctuations, join them again to form the sentence.
    dummy_sentence = [c for c in sentence if c not in punctuations]
    print(color.BLUE,"Character list without punctuations:\n", color.END,dummy_sentence)
    sentence = "".join(dummy_sentence).strip().split(" ")
    for i in sentence:
        words.append(i)

Punctuation: !"#$%&'()*+,-./:;<=>?@[\]^_`{|}~
--------------------
Type of char_string: <class 'str'>
--------------------
Character at position 3 of word_string is : 'l'
--------------------
Single Line string:  Hello World.
--------------------
--------------------
 Sentence:   Lorem ipsum dolor sit amet, consectetur adipiscing elit. 
 Character list without punctuations:
  ['L', 'o', 'r', 'e', 'm', ' ', 'i', 'p', 's', 'u', 'm', ' ', 'd', 'o', 'l', 'o', 'r', ' ', 's', 'i', 't', ' ', 'a', 'm', 'e', 't', ' ', 'c', 'o', 'n', 's', 'e', 'c', 't', 'e', 't', 'u', 'r', ' ', 'a', 'd', 'i', 'p', 'i', 's', 'c', 'i', 'n', 'g', ' ', 'e', 'l', 'i', 't', ' ']
--------------------
 Sentence:   Duis laoreet tempor tempus. 
 Character list without punctuations:
  ['D', 'u', 'i', 's', ' ', 'l', 'a', 'o', 'r', 'e', 'e', 't', ' ', 't', 'e', 'm', 'p', 'o', 'r', ' ', 't', 'e', 'm', 'p', 'u', 's', ' ']
--------------------
 Sentence:   Nulla vel mi vel nibh congue interdum. 
 Character list without punctuat

In [ ]:
print(*sentences, sep="\n")

Lorem ipsum dolor sit amet, consectetur adipiscing elit. 
Duis laoreet tempor tempus. 
Nulla vel mi vel nibh congue interdum. 
Pellentesque elit nisl, vulputate quis erat eget, eleifend maximus lorem. 
Praesent placerat sem in ipsum ultrices, sed tincidunt quam consectetur. 
Donec vel nisi et lorem porta posuere vitae vitae justo. 
Pellentesque rhoncus scelerisque ex, facilisis consequat leo aliquam vel. 
Vestibulum facilisis, nibh sit amet luctus dapibus, tellus magna euismod nisi, in condimentum mi odio sed ligula. 
Aliquam purus est, imperdiet ac porttitor ac, eleifend in odio.


In [ ]:
print(words)

['Lorem', 'ipsum', 'dolor', 'sit', 'amet', 'consectetur', 'adipiscing', 'elit', 'Duis', 'laoreet', 'tempor', 'tempus', 'Nulla', 'vel', 'mi', 'vel', 'nibh', 'congue', 'interdum', 'Pellentesque', 'elit', 'nisl', 'vulputate', 'quis', 'erat', 'eget', 'eleifend', 'maximus', 'lorem', 'Praesent', 'placerat', 'sem', 'in', 'ipsum', 'ultrices', 'sed', 'tincidunt', 'quam', 'consectetur', 'Donec', 'vel', 'nisi', 'et', 'lorem', 'porta', 'posuere', 'vitae', 'vitae', 'justo', 'Pellentesque', 'rhoncus', 'scelerisque', 'ex', 'facilisis', 'consequat', 'leo', 'aliquam', 'vel', 'Vestibulum', 'facilisis', 'nibh', 'sit', 'amet', 'luctus', 'dapibus', 'tellus', 'magna', 'euismod', 'nisi', 'in', 'condimentum', 'mi', 'odio', 'sed', 'ligula', 'Aliquam', 'purus', 'est', 'imperdiet', 'ac', 'porttitor', 'ac', 'eleifend', 'in', 'odio']


In [ ]:
print("-"*20)
print(color.BOLD,"All words in lower case:\n",color.END,[i.lower() for i in words])
print("-"*20)
print("All words in upper case:\n",[i.upper() for i in words])
print("-"*20)
print("All words in title case:\n",[i.title() for i in words])
print("-"*20)
print("All words in swapcase:\n",[i.swapcase() for i in words])
print("-"*20)
print("All words in that startswith 'sc':\n",[i for i in words if i.startswith("sc")])
print("-"*20)
print("All words in that endswith 'am':\n",[i for i in words if i.endswith("am")])
print("-"*20)

--------------------
 All words in lower case:
  ['lorem', 'ipsum', 'dolor', 'sit', 'amet', 'consectetur', 'adipiscing', 'elit', 'duis', 'laoreet', 'tempor', 'tempus', 'nulla', 'vel', 'mi', 'vel', 'nibh', 'congue', 'interdum', 'pellentesque', 'elit', 'nisl', 'vulputate', 'quis', 'erat', 'eget', 'eleifend', 'maximus', 'lorem', 'praesent', 'placerat', 'sem', 'in', 'ipsum', 'ultrices', 'sed', 'tincidunt', 'quam', 'consectetur', 'donec', 'vel', 'nisi', 'et', 'lorem', 'porta', 'posuere', 'vitae', 'vitae', 'justo', 'pellentesque', 'rhoncus', 'scelerisque', 'ex', 'facilisis', 'consequat', 'leo', 'aliquam', 'vel', 'vestibulum', 'facilisis', 'nibh', 'sit', 'amet', 'luctus', 'dapibus', 'tellus', 'magna', 'euismod', 'nisi', 'in', 'condimentum', 'mi', 'odio', 'sed', 'ligula', 'aliquam', 'purus', 'est', 'imperdiet', 'ac', 'porttitor', 'ac', 'eleifend', 'in', 'odio']
--------------------
All words in upper case:
 ['LOREM', 'IPSUM', 'DOLOR', 'SIT', 'AMET', 'CONSECTETUR', 'ADIPISCING', 'ELIT', 'DUIS',

## Python Flow Control (if ... else)
---
- `if ... else` statements are used for flow control in python
- `elif` is used for the else if cases
- Python is a indented language, i.e. it does not use any `{}`. It only uses indentations and a colon `:`.

In [ ]:
# Let's look into a very simple for loop example
days = ["Monday", "Tuesday", "Wednesday", "Thursday", "Friday", "Saturday", "Sunday"]
for day in days:
    if day=="Sunday":
        print("It's",day,". Time to sleep again.")
    elif day=="Saturday":
        print("Today is",day,". Do all the leftout work.")
    else:
        print("Hey it's",day,". Time to work.")

Hey it's Monday . Time to work.
Hey it's Tuesday . Time to work.
Hey it's Wednesday . Time to work.
Hey it's Thursday . Time to work.
Hey it's Friday . Time to work.
Today is Saturday . Do all the leftout work.
It's Sunday . Time to sleep again.


## `while` loop in Python via Guessing game
----
- As in case of any other programming language, while loops are used when we are unaware of the number of iterations that should be run, but we know the condition when to stop.
- Below is an example code to guess a number, using `while` loop



In [ ]:
number = 89
guess = None
while guess!=number:
    guess = int(input("Enter a number: "))
    if guess==number:
        print("Yes you guessed it right. It's {}".format(guess))
    else:
        if guess<number:
            print("No it's not {}. Try guessing a larger number.".format(guess))
        else:
            print("No it's not {}. Try guessing a smaller number.".format(guess))

Enter a number: 89
Yes you guessed it right. It's 89


## Python Functions
---
- Functions are blocks of codes that can be reused to perform some action.
- Can be used to modularize codes
- Function BLock begins with the keyword `def` followed by the function name and parenthesis `()`.
- Parameters or arguments to the function are to passed within these parenthesis
- Docstrings can be used to describe the function.
- Code inside the function block has to be indented
- `return` is used to return the function output  
- We were using a function to return the fibonacci sequence. Now, we can provide the number upto which we want the sequence and then call the function `fibonacci` to calculate.  
- We'll also see a function to calculate factorial of a number. The recursive definition of factorial being:  
$fac(n)$ = $n*fac(n-1)$  
with $fac(1)$ = 1 and $n>=2$  
- The arguments passed to the functions can be:  
    - **Default arguments** : assign a default value to the argument. Need to make sure that `non-default argument follows default argument`.
    - **Keyword arguments** : With keyword arguments in python, we can change the order of passing the arguments without any consequences.
    - **Arbitrary Arguments** : This can be used, when the number to arguments to expect to a function are unknown.



In [ ]:
# Example of default argument
fib_seq = [0,1]
fib_ratio = []
def fibonacci(n=10):
    if n<0:
        return 0
    elif n<len(fib_seq):
        return fib_seq[n-1], fib_ratio[n-3]
    else:
        for i in range(len(fib_seq),n):
            fib_seq.append(fib_seq[-1]+fib_seq[-2])
            fib_ratio.append(fib_seq[-1]/fib_seq[-2])

        return fib_seq[n-1], fib_ratio[n-3]


print(fibonacci(5))
print(fibonacci(n=10))

(3, 1.5)
(34, 1.619047619047619)


In [ ]:
# Example of keyword argument
def add(a,b=2):
    return a+b

# These work
print(color.GREEN,"add(a=3,b=4): ",color.END,add(a=3,b=4),sep="")
print(color.GREEN,"add(a=3): ",color.END,add(a=3),sep="")
try:
    print(color.GREEN,"add()",color.END,add(),sep="")
except Exception as e:
    print(color.GREEN,"add(a), raises exception: ",color.END,e,sep="")

add(a=3,b=4): 7
add(a=3): 5
add(a), raises exception: add() missing 1 required positional argument: 'a'


<div class="alert alert-block alert-info">
Python function arguments should follow the order given below:  
<ul>
    <li> Formal arguments </li>
    <li> *args : collects extra positional arguments as a tuple</li>
    <li> Keyword arguments </li>
    <li> **kwargs : collects extra keyword arguments as a dictionary. </li>
    <li> * and ** can be used to unpack function arguments from sequences and dictionaries </li>
</ul>

In [ ]:
def names(n, *args, n1="David", **kwargs):
    print(color.RED,"n = ",color.END, n, sep="")
    print(color.RED,"type(args) = ",color.END,type(args), sep="")
    print(color.RED,"args = ",color.END, args, sep="")
    print(color.RED,"n1 = ",color.END, n1, sep="")
    print(color.RED,"type(kwargs) = ",color.END, type(kwargs), sep="")
    print(color.RED,"kwargs = ",color.END, kwargs, sep="")

names("James","John", "Robert", name1="Michael", name2="William")

n = James
type(args) = <class 'tuple'>
args = ('John', 'Robert')
n1 = David
type(kwargs) = <class 'dict'>
kwargs = {'name1': 'Michael', 'name2': 'William'}


In [ ]:
def factorial(n=1):
    """ Calculates the factorial of a number n
        if n==0 or n==1:
            return 1
        else: #calculates recursively
            n*factorial(n-1)
        :param n: (int) the number of which to calculate the factorial
        :output : (int) returns the factorial of the number
    """
    if n<0:
        return "Give a positive number as input"
    elif n==0 or n==1:
        return 1
    else:
        return n*factorial(n-1)

print(color.BLUE,"factorial(7):",color.END,factorial(7))

 factorial(7):  5040


## Collatz Conjecture
---
[Source](https://en.wikipedia.org/wiki/Collatz_conjecture)
<table>
<tr>
<td>

The Collatz conjecture is a conjecture in mathematics that concerns a sequence defined as follows:
<ul>
<li>Start with any positive integer `n`. </li>
<li>Then each term is obtained from the previous term as follows: </li>
    <ul>
    <li>if the previous term is even, the next term is one half of the previous term. (i.e. `n/2` ) </li>
    <li> If the previous term is odd, the next term is 3 times the previous term plus 1 (i.e. `3n+1` ).
    </ul>
<li><b>The conjecture is that no matter what value of `n`, the sequence will always reach 1.</b></li>
</ul>
</td>

<td>
<img src="https://imgs.xkcd.com/comics/collatz_conjecture.png" alt="https://imgs.xkcd.com/comics/collatz_conjecture.png" width=500></img>
</td>

</tr>
</table>

In [ ]:
def collatz_conjecture(n=100, verbose=False):
    temp = n
    count = 0
    while temp!=1:
        if temp%2==0:
            temp = temp//2
        else:
            temp = 3*temp+1
        count += 1
        if verbose:
            print("At count {}, the number is {}.".format(count, temp))
    print("Reached conjecture at count:",count)

print(color.GREEN, "type(collatz_conjecture): ", color.END, type(collatz_conjecture))
for i in range(100,151,5):
    print(color.GREEN,"number =",color.END,i,end=" ")
    collatz_conjecture(i)

 type(collatz_conjecture):   <class 'function'>
 number =  100 Reached conjecture at count: 25
 number =  105 Reached conjecture at count: 38
 number =  110 Reached conjecture at count: 113
 number =  115 Reached conjecture at count: 33
 number =  120 Reached conjecture at count: 20
 number =  125 Reached conjecture at count: 108
 number =  130 Reached conjecture at count: 28
 number =  135 Reached conjecture at count: 41
 number =  140 Reached conjecture at count: 15
 number =  145 Reached conjecture at count: 116
 number =  150 Reached conjecture at count: 15


## Lambda Functions and Functional Programming
---
- There are different paradigms of programming that are present, one of which is **Functional Programming**.  
- Python `lambda` functions are a gateway to the Functional programming nature of python.
- `lambda` functions are one-line functions that performs some operations without requiring the `def` statement.
- `map` and `filter` are other helper functions that can be used adjunct to `lambda` functions for fast computations

In [ ]:
square= lambda x: x*x if isinstance(x,int) else [i*i for i in x]

print(color.BLUE,"type(square):",color.END,type(square))
print(color.BLUE,"square(2)):",color.END,square(2))
print(color.BLUE,"square(range(10)):",color.END,square(range(10)))
print(color.BLUE,"square(list(range(10,20))):",color.END,square(list(range(10,20))))

 type(square):  <class 'function'>
 square(2)):  4
 square(range(10)):  [0, 1, 4, 9, 16, 25, 36, 49, 64, 81]
 square(list(range(10,20))):  [100, 121, 144, 169, 196, 225, 256, 289, 324, 361]


### `map` in Python

In [ ]:
in_string = "1 2 3 4 5 6 7 8 9"
# If we want to convert the above string to a list of integers
# we would first have to split the string and then convert each
# element back to integer
# we can use list comprehension
print(color.RED,"Using List Comprehension:",color.END,[int(i) for i in in_string.split()])
# there is a more readable way using map
in_list = list(map(int, in_string.split()))
print(color.RED,"Using `map`:",color.END,in_list)
print(color.RED,"square using map:",color.END,end="")
print(list(map(square, in_list)))

 Using List Comprehension:  [1, 2, 3, 4, 5, 6, 7, 8, 9]
 Using `map`:  [1, 2, 3, 4, 5, 6, 7, 8, 9]
 square using map: [1, 4, 9, 16, 25, 36, 49, 64, 81]


### `filter` in Python

In [ ]:
# example of filters
even = lambda x: x%2==0
print(color.GREEN,"Example of using filter to store only the values from 0 to 21 that are even to a list:",color.END, sep="")
print(list(filter(even, range(21))))

Example of using filter to store only the values from 0 to 21 that are even to a list:
[0, 2, 4, 6, 8, 10, 12, 14, 16, 18, 20]


## Python File Handling
----
- Python offers a very interface to open, read, write and append to files of any kind.
- Python has a syntax-like `with open(...) as ... :` to access the files. (best-practice)  
`open()` takes a path-like object to the file we want to open and the `mode` in which the file needs to be opened.  
`mode`s available are:  
    - "r" : read-only mode
    - "w" : write-only mode
    - "a" : append-only mode
    - "rb" : read-binary mode
    - "wb" : write-binary mode
    - "r+" : read/write mode
    - "a+" : append and read mode



In [ ]:
# The way to create a new file and write to it is as follows:
with open("dummy.txt","w") as f:
    for i in range(10):
        f.write("This is line number {}.\n".format(i))

# Now let's look at the file content by downloading it.

In [ ]:
# The way to read a file is as follows
with open("dummy.txt","r") as f:
    print(color.BLUE,"Name of the file opened: ",color.END, f.name,
          color.BLUE," with mode: ",color.END,f.mode, sep="")
    for line in f.readlines():
        print(line.strip("\n"))

Name of the file opened: dummy.txt with mode: r
This is line number 0.
This is line number 1.
This is line number 2.
This is line number 3.
This is line number 4.
This is line number 5.
This is line number 6.
This is line number 7.
This is line number 8.
This is line number 9.


## Python Modules
---
- Along with the built-in functions, Python has loads of built-in modules to help us with computations, and other programming tasks.
- Some of the most useful and commonly used built-in modules are:  
    - `os` : useful interaction with OS
    - `sys` : useful for system calls
    - `time` : dealing with time data
    - `math` : for mathematical operations
    - `datetime` : for datetime data parsing
    - `random` : for random number operations
    - `csv` : to read csv files
    - `re` : for regular expression works
    - `string` : useful for string operations
    - `urllib` : useful for using web page parsing
    - `collections` : useful for computations
    - `itertools` : useful for iterative computations
    - `statistics` : for statistical functions like mean, median, mode, etc.  
- In order to import a module in python, we use the `import` keyword. E.g. if we want to import `math` module into our code, we use the command `import math`
- It is conventional to define all the imports at the top of the python scripts or codes, for better readability.

### Glimpse of `math` module
---
Python's `math` module has some useful functions for mathematical calculations. Some of them are:  
- constants: `math.pi` ($\pi$), `math.e` ($e$), `math.tau` ($\tau = 2*\pi$)
- logarithms: `math.log()` ($ln$), `math.log10` ($log_{10}$) and `math.log2` ($log_{2}$)
- arithmetic functions: `factorial()`, `ceil()`, `floor()`, `trunc()`, `pow()`, `exp()`, `gcd()`, `sqrt()`
- trigonometric functions: `sin`, `cos`, `tan`, `asin`, `acos` and `atan`
- number theory and algebraic functions: `comb`, `perm`, `isqrt`, `prod`, `dist`, `hypot`, but these are only for python version 3.8+

- There is a separate library for complex number mathematics, called `cmath`, with similar functions for complex arithmetic

In [ ]:
# Math module
import math
# Constants
print(color.RED,"Value of Pi: ",color.END, math.pi)
print(color.RED,"Value of e: ",color.END, math.e)
print(color.RED,"Value of tau (2*Pi) : ",color.END, math.tau)

# Let's create a function to calculate the area and circumference of a circle
def area_circumference(radius=1.0):
    area = math.pi*radius*radius
    circumference = 2*math.pi*radius
    return area, circumference

print(color.RED,"Value of area_circumference(5): ",color.END, *area_circumference(5))

 Value of Pi:   3.141592653589793
 Value of e:   2.718281828459045
 Value of tau (2*Pi) :   6.283185307179586
 Value of area_circumference(5):   78.53981633974483 31.41592653589793


In [ ]:
#-----------------------
# Logarithm functions
#-----------------------
print("""#-----------------------
# Logarithmic Functions
#-----------------------""")
print(color.BLUE,"math.log(math.e): ", color.END, math.log(math.e),sep="")
print(color.BLUE,"math.log(10): ", color.END, math.log(10),sep="")
print(color.BLUE,"math.log(10,10): ", color.END, math.log(10,10),sep="")
print(color.BLUE,"math.log10(5): ", color.END, math.log10(5),sep="")
print(color.BLUE,"math.log2(8): ", color.END, math.log2(8),sep="")

#-----------------------
# Arithmetic functions
#-----------------------
print("""#-----------------------
# Arithmetic Functions
#-----------------------""")
print(color.BLUE,"math.factorial(5): ", color.END, math.factorial(5),sep="")
print(color.BLUE,"math.gcd(17,51): ", color.END, math.gcd(17,51),sep="")
# calculation of lcm is not present in math module
print(color.BLUE,"lcm(12,56) using (12*56)/math.gcd(12,56): ", color.END, (12*56)/math.gcd(12,56),sep="")
print(color.BLUE,"math.trunc(math.pow(5,2)): ", color.END, math.trunc(math.pow(5,2)),sep="")
print(color.BLUE,"math.ceil(5.3): ", color.END, math.ceil(5.3),sep="")
print(color.BLUE,"math.floor(5.3): ", color.END, math.floor(5.3),sep="")
print(color.BLUE,"math.sqrt(5.3): ", color.END, math.sqrt(5.3),sep="")
try:
    print(color.BLUE,"math.isqrt(5.3): ", color.END, math.isqrt(5.3),sep="")
except Exception as e:
    print(e)

#-----------------------
# Trigonometric Functions
#-----------------------
print("""#-----------------------
# Trigonometric Functions
#-----------------------""")
# use math.radians to convert from degrees to radians
# and math.degrees to convert from radians back to degrees
print(color.BLUE,"math.sin(90): ", color.END, math.sin(math.radians(90)),sep="")
print(color.BLUE,"math.cos(90): ", color.END, math.cos(math.radians(90)),sep="")
print(color.BLUE,"math.tan(90): ", color.END, math.tan(math.radians(90)),sep="")
print(color.BLUE,"math.asin(1): ", color.END, math.degrees(math.asin(1)),sep="")
print(color.BLUE,"math.acos(1): ", color.END, math.degrees(math.acos(1)),sep="")
print(color.BLUE,"math.atan(1): ", color.END, math.degrees(math.atan(1)),sep="")

#-----------------------
# Logarithmic Functions
#-----------------------
math.log(math.e): 1.0
math.log(10): 2.302585092994046
math.log(10,10): 1.0
math.log10(5): 0.6989700043360189
math.log2(8): 3.0
#-----------------------
# Arithmetic Functions
#-----------------------
math.factorial(5): 120
math.gcd(17,51): 17
lcm(12,56) using (12*56)/math.gcd(12,56): 168.0
math.trunc(math.pow(5,2)): 25
math.ceil(5.3): 6
math.floor(5.3): 5
math.sqrt(5.3): 2.3021728866442674
module 'math' has no attribute 'isqrt'
#-----------------------
# Trigonometric Functions
#-----------------------
math.sin(90): 1.0
math.cos(90): 6.123233995736766e-17
math.tan(90): 1.633123935319537e+16
math.asin(1): 90.0
math.acos(1): 0.0
math.atan(1): 45.0


In [ ]:
print("""#-----------------------
# Complex Math Functions
#-----------------------""")
import cmath
print(color.BLUE,"cmath.log(2+3j): ", color.END, cmath.log(2+3j),sep="")
print(color.BLUE,"cmath.log10(2+3j): ", color.END, cmath.log10(2+3j),sep="")
print(color.BLUE,"cmath.exp(2+3j): ", color.END, cmath.exp(2+3j),sep="")
print(color.BLUE,"cmath.polar(2+3j): ", color.END, cmath.polar(2+3j),sep="")
print(color.BLUE,"cmath.cos(2+3j): ", color.END, cmath.cos(2+3j),sep="")
# etc ...

#-----------------------
# Complex Math Functions
#-----------------------
cmath.log(2+3j): (1.2824746787307684+0.982793723247329j)
cmath.log10(2+3j): (0.5569716761534184+0.42682189085546657j)
cmath.exp(2+3j): (-7.315110094901103+1.0427436562359045j)
cmath.polar(2+3j): (3.605551275463989, 0.982793723247329)
cmath.cos(2+3j): (-4.189625690968807-9.109227893755337j)


## Case Study
---
#### Describe some statistics of the famouse IRIS Dataset
---
- The IRIS dataset [Source](https://archive.ics.uci.edu/ml/machine-learning-databases/iris/iris.data) has about 150 instances of 3 classes of flowers, with 50 instances of each.
- Each class refers to a type of the IRIS Plant.
- Attribute Information (what each of these columns mean):  
    1. sepal length (in cm)
    2. sepal width (in cm)
    3. petal length (in cm)
    4. petal width (in cm)
    5. class:  
        - Iris Setosa
        - Iris Versicolour
        - Iris Virginica

In [ ]:
# Here we look into a way of downloading the iris dataset from the online source
data_url="https://archive.ics.uci.edu/ml/machine-learning-databases/iris/iris.data"
import urllib.request
import time

start= time.time()

urllib.request.urlretrieve(data_url,"iris_data.csv")

# Next we look into how we want to read the file and store the information for use
# I would like to store it in a dictionary, for the conveniene that different
# columns can then be used as keys for us to calculate things later
data_dict = dict()
import csv
with open("iris_data.csv","r") as f:
    csv_reader = csv.reader(f, delimiter=",")
    for row in csv_reader:
        if len(row)>0:
            if row[4] in data_dict:
                data_dict[row[4]]["sepal_length"].append(float(row[0]))
                data_dict[row[4]]["sepal_width"].append(float(row[1]))
                data_dict[row[4]]["petal_length"].append(float(row[2]))
                data_dict[row[4]]["petal_width"].append(float(row[3]))
            else:
                data_dict[row[4]] = {"sepal_length": [float(row[0])],
                                    "sepal_width": [float(row[1])],
                                    "petal_length":[float(row[2])],
                                    "petal_width": [float(row[3])], }
print(data_dict.keys())
print(data_dict.get('Iris-setosa'))

dict_keys(['Iris-setosa', 'Iris-versicolor', 'Iris-virginica'])
{'sepal_length': [5.1, 4.9, 4.7, 4.6, 5.0, 5.4, 4.6, 5.0, 4.4, 4.9, 5.4, 4.8, 4.8, 4.3, 5.8, 5.7, 5.4, 5.1, 5.7, 5.1, 5.4, 5.1, 4.6, 5.1, 4.8, 5.0, 5.0, 5.2, 5.2, 4.7, 4.8, 5.4, 5.2, 5.5, 4.9, 5.0, 5.5, 4.9, 4.4, 5.1, 5.0, 4.5, 4.4, 5.0, 5.1, 4.8, 5.1, 4.6, 5.3, 5.0], 'sepal_width': [3.5, 3.0, 3.2, 3.1, 3.6, 3.9, 3.4, 3.4, 2.9, 3.1, 3.7, 3.4, 3.0, 3.0, 4.0, 4.4, 3.9, 3.5, 3.8, 3.8, 3.4, 3.7, 3.6, 3.3, 3.4, 3.0, 3.4, 3.5, 3.4, 3.2, 3.1, 3.4, 4.1, 4.2, 3.1, 3.2, 3.5, 3.1, 3.0, 3.4, 3.5, 2.3, 3.2, 3.5, 3.8, 3.0, 3.8, 3.2, 3.7, 3.3], 'petal_length': [1.4, 1.4, 1.3, 1.5, 1.4, 1.7, 1.4, 1.5, 1.4, 1.5, 1.5, 1.6, 1.4, 1.1, 1.2, 1.5, 1.3, 1.4, 1.7, 1.5, 1.7, 1.5, 1.0, 1.7, 1.9, 1.6, 1.6, 1.5, 1.4, 1.6, 1.6, 1.5, 1.5, 1.4, 1.5, 1.2, 1.3, 1.5, 1.3, 1.5, 1.3, 1.3, 1.3, 1.6, 1.9, 1.4, 1.6, 1.4, 1.5, 1.4], 'petal_width': [0.2, 0.2, 0.2, 0.2, 0.2, 0.4, 0.3, 0.2, 0.2, 0.1, 0.2, 0.2, 0.1, 0.1, 0.2, 0.4, 0.4, 0.3, 0.3, 0.3, 0.2, 0.4, 0.2, 0

In [ ]:
## Now that we have loaded the dataset into a dictionary we can now use maths to look into soem features statistics
# Let's look at the minimum, maximum, means, standard deviation of each of these features for each class.
import statistics as stat
# python also has a built in statistics module

classes = data_dict.keys()
overall_stats={
"sepal_length": {"min":[], "max":[], "mean":[], "std": []},
"sepal_width": {"min":[], "max":[], "mean":[], "std": []},
"petal_length": {"min":[], "max":[], "mean":[], "std": []},
"petal_width": {"min":[], "max":[], "mean":[], "std": []}
}

for class_ in classes:
    print(color.BLUE,"For iris class: ",color.END,class_,sep="")
    attributes = data_dict[class_].keys()
    print("\t".join(["attribute","min","max","mean","std"]))
    for i,attr in enumerate(attributes):
        min_ = min(data_dict[class_][attr])
        max_ = max(data_dict[class_][attr])
        mean_ = stat.mean(data_dict[class_][attr])
        std_ = stat.stdev(data_dict[class_][attr])

        overall_stats[attr]["min"].append(min_)
        overall_stats[attr]["max"].append(max_)
        overall_stats[attr]["mean"].append(mean_)
        overall_stats[attr]["std"].append(std_)

        print("{}\t{}\t{}\t{:.2f}\t{:.2f}".format(attr, min_, max_, mean_, std_))


print(color.BLUE,"""-------------------------------------------
\t\tOverall Statistics
-------------------------------------------""",color.END, sep="")
print("\t".join(["attribute","min","max","mean","std"]))
for i,attr in enumerate(attributes):
    print("{}\t{:.2f}\t{:.2f}\t{:.2f}\t{:.2f}".format(attr,
                                              min(overall_stats[attr]["min"]),
                                              max(overall_stats[attr]["max"]),
                                              stat.mean(overall_stats[attr]["mean"]),
                                              stat.mean(overall_stats[attr]["std"])))

print(color.GREEN,"Time taken for this execution is :", color.END, round(time.time()-start,2)," sec", sep="")

For iris class: Iris-setosa
attribute	min	max	mean	std
sepal_length	4.3	5.8	5.01	0.35
sepal_width	2.3	4.4	3.42	0.38
petal_length	1.0	1.9	1.46	0.17
petal_width	0.1	0.6	0.24	0.11
For iris class: Iris-versicolor
attribute	min	max	mean	std
sepal_length	4.9	7.0	5.94	0.52
sepal_width	2.0	3.4	2.77	0.31
petal_length	3.0	5.1	4.26	0.47
petal_width	1.0	1.8	1.33	0.20
For iris class: Iris-virginica
attribute	min	max	mean	std
sepal_length	4.9	7.9	6.59	0.64
sepal_width	2.2	3.8	2.97	0.32
petal_length	4.5	6.9	5.55	0.55
petal_width	1.4	2.5	2.03	0.27
-------------------------------------------
		Overall Statistics
-------------------------------------------
attribute	min	max	mean	std
sepal_length	4.30	7.90	5.84	0.50
sepal_width	2.00	4.40	3.05	0.34
petal_length	1.00	6.90	3.76	0.40
petal_width	0.10	2.50	1.20	0.19
Time taken for this execution is :0.2 sec


# Python Classes and Objects
---
We had seen before how to use Lists as a Stack. Below we represent a more formal way to implement stacks.

In [ ]:
class Stack:
    """ Stack (LIFO) class.
        Options: - add(x) : add element x to the stack instance
                 - pop() : pop the last element of the stack
                 - insert(x,pos) : insert element x, at position pos
                 - remove(x) : remove one instance of from stack

        Usage: - s = Stack([1,2,3]) : if we want to initialize a stack with list
               - s.add(7) : add 7 to the end of the stack
               - s.pop() : to pop the last element
    """
    def __init__(self, x=[]):
        if not isinstance(x,list):
            self.stack = list(x)
        else:
            self.stack = x

    def add(self, x):
        if not isinstance(x,list):
            self.stack.append(x)
        else:
            self.stack.extend(x)

    def pop(self):
        return self.stack.pop()

    def insert(self, x, pos=None):
        if pos==None:
            self.stack.insert(len(self.stack), x)
        else:
            self.stack.insert(pos, x)

    def remove(self, x):
        self.stack.remove(x)

    def __str__(self):
        return "{}".format(self.stack)

In [ ]:
s = Stack()
print(s)
s.add([1,2,3,45,50,10])
print(s)
pop = s.pop()
print(pop, s)

[]
[1, 2, 3, 45, 50, 10]
10 [1, 2, 3, 45, 50]
